02-microarray/pathway-analysis_microarray_03_gsva.Rmd

```diff
@@ -0,0 +1,428 @@
+
+Let's see a preview of `mapped_df`.
+
+```{r}
+head(mapped_df)
+```
+
+We will want to keep in mind that GSVA requires that data is in a matrix with the gene identifiers as rownames.
+In order to successfully turn our data frame into a matrix, we will need to ensure that we do not have any duplicate gene identifiers.
+
+Let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.
+
+```{r}
+any(duplicated(mapped_df$entrez_id))
+```
+
+Looks like we do have duplicated Entrez IDs.
+Let's find out which Entrez IDs have been duplicated.
+
+```{r}
+dup_entrez_ids <- mapped_df %>%
+  dplyr::filter(duplicated(entrez_id)) %>%
+  dplyr::pull(entrez_id)
+
+dup_entrez_ids
+```
+
+Now let's take a look at the rows associated with the duplicated Entrez IDs.
+
+```{r}
+mapped_df %>%
+  dplyr::filter(entrez_id %in% dup_entrez_ids)
+```
+
+We can see that the associated values vary for each row.
+
+As we mentioned earlier, we will want to remove duplicated gene identifiers in preparation for the GSVA steps later, so let's keep the Entrez IDs associated with the maximum expression value per sample.
+Note that using the maximum expression value to collapse duplicate gene identifiers is just one of the expression values that can be used here.
+See the [Gene Set Enrichment Analysis (GSEA) User guide](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html) for more information on other expression values that can be used [@GSEA-user-guide]. 
+
+We are removing values for three gene identifiers here, so it is unlikely to have a considerable impact on our results.
+
+```{r}
+max_dup_df <- mapped_df %>%
+  # Filter to include only the rows associated with the duplicate gene identifiers
+  dplyr::filter(entrez_id %in% dup_entrez_ids) %>%
+  # Group by Entrez IDs
+  dplyr::group_by(entrez_id) %>%
+  # Get the maximum expression value using all values associated with the grouping variable
+  dplyr::summarize_all(max)
+
+max_dup_df
+```
+Looks like we were able to successfully get rid of the duplicate gene identifiers and keep the observations with the maximum expression value!
+
+Now let's combine the rest of the mapped data!
+
+```{r}
+filtered_mapped_df <- mapped_df %>%
+  dplyr::filter(
+    # First let's get the data associated with the gene identifiers that aren't duplicated
+    !entrez_id %in% dup_entrez_ids |
+      # Now let's get the data associated with the duplicated gene identifiers that we
+      # collapsed in `max_dup_df`
+      entrez_id %in% max_dup_df$entrez_id &
+        Ensembl %in% max_dup_df$Ensembl
+  )
+```
+
+Let's check to see that we removed the duplicate Entrez IDs from the rest of the mapped data.
+
+```{r}
+filtered_mapped_df %>%
+  # Let's sort by the Entrez IDs we know were duplicated
+  dplyr::arrange(dplyr::desc(entrez_id %in% dup_entrez_ids)) %>%
```

Comment:
If we are just doing a preview, I think this can be a lot simpler: 
```suggestion
  dplyr::filter(entrez_id %in% dup_entrez_ids) %>%
```

Reply:
I changed this to `head()` now that we arrange by the `dup_entrez_ids` in the previous step.